keps/sig-node/2000-graceful-node-shutdown/README.md

```diff
@@ -0,0 +1,843 @@
+    node shutdowns
+
+### Background on Linux Shutdown
+
+In the context of this KEP, shutdown is referred to as shutdown of the
+underlying machine. On most linux distros shutdown can be initiated via a
+variety of methods for example:
+
+1. `shutdown -h now`
+2. `shutdown -h +30` `#schedule a delayed shutdown in 30mins`
+3. `systemctl poweroff`
+4. Physically pressing the power button on the machine
+5. If a machine is a VM, the underlying hypervisor can press the “virtual”
+   power button
+6. For a cloud instance, stopping the instance via Cloud API, e.g. via `gcloud
+   compute instances stop`. Depending on the cloud provider, this may result in
+   virtual power button press by the underlying hypervisor.
+
+
+Some of these cases will involve the machine receiving an ACPI event to change
+the power state. The machine can go from G0 (working state) to G2 (Soft Off)
+and finally to G3 (Off) [more info on ACPI
+states](https://en.wikipedia.org/wiki/Advanced_Configuration_and_Power_Interface).
+On Linux, prior to shutdown usually a system demon will listen to these events
+and perform some series of actions prior to userspace calling the
+[reboot(2)](https://man7.org/linux/man-pages/man2/reboot.2.html) systemcall
+with `LINUX_REBOOT_CMD_POWER_OFF` or `LINUX_REBOOT_CMD_HALT` to actually
+shutdown the machine.
+
+Historically, ACPI events were often handled by the
+[acpid](https://wiki.archlinux.org/index.php/Acpid) daemon which uses a variety
+of mechanisms to watch ACPI events (i.e. reading `/proc/acpi/event` or
+`/dev/input/eventX` to react to power button presses). However, in most modern
+linux distros today,
+[systemd-logind](https://www.freedesktop.org/wiki/Software/systemd/logind/)
+has taken over as the main component [reacting to ACPI
+events](https://wiki.archlinux.org/index.php/Power_management#ACPI_events) and
+initiating shutdown of the machine. On a system with
+systemd-logind, for example, a trigger of the power button will
+result in the systemd target
+[poweroff](https://www.freedesktop.org/software/systemd/man/systemd-halt.service.html)
+being run (see
+[HandlePowerKey](https://www.freedesktop.org/software/systemd/man/logind.conf.html),
+which will terminate all the systemd services running on the machine and
+eventually shut it down. However, in the context of kubernetes, systemd is not
+aware of the pods and containers running on the machine and systemd will simply
+kill them as regular linux processes.
+
+### Background on Inhibitors
+
+`systemd-logind` provides the ability for applications to delay shutdown and
+perform some series of actions before the shutdown completes through a
+mechanism called ["Inhibitor
+Locks".](https://www.freedesktop.org/wiki/Software/systemd/inhibit/)
+Applications can request to delay shutdown by taking an inhibitor lock by
+sending messages to logind over dbus. Applications can request up to
+`InhibitDelayMaxSec` (a setting configured in logind.conf`) for delay based
```

Comment:
Niit-pick: I think there is a missing ` before logind.conf that makes the formatting wrong (because there is one at the end of logind.conf, closing)

Reply:
Thanks, fixed.